compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/ClassGenerator.kt

```diff
@@ -106,30 +113,97 @@ class ClassGenerator(
-                type.toIrType() as? IrSimpleType ?: error("Value class underlying type is not a simple type: $classDescriptor")
-            }
+            if (ktClassOrObject is KtClassOrObject &&
+                (classDescriptor.isInlineClass() || classDescriptor.isValue && classDescriptor.modality == Modality.SEALED)
+            ) {
+                irClass.valueClassRepresentation =
+                    if (irClass.modality == Modality.SEALED)
+                        InlineClassRepresentation(Name.identifier("\$value"), context.irBuiltIns.anyNType as IrSimpleType)
+                    else {
+                        val representation = classDescriptor.valueClassRepresentation
+                            ?: error("Unknown representation for inline class: $classDescriptor")
+                        representation.mapUnderlyingType { type ->
+                            val irType = type.toIrType() as? IrSimpleType
+                                ?: error("Inline class underlying type is not a simple type: $classDescriptor")
+
+                            if (classDescriptor.isInlineClass() && classDescriptor.getSuperClassOrAny().isSealedInlineClass() &&
+                                type.isPrimitiveNumberType()
+                            ) {
+                                irType.makeNullable() as? IrSimpleType
+                                    ?: error("Inline class underlying type is not a simple type: $classDescriptor")
+                            } else irType
+                        }
+                    }
+
+                val isSealedInlineSubclassOfSealedInlineClass =
+                    classDescriptor.getSuperClassOrAny().isInlineClass() && classDescriptor.modality == Modality.SEALED
 
-            if (irClass.isSingleFieldValueClass && ktClassOrObject is KtClassOrObject) {
-                generateAdditionalMembersForSingleFieldValueClasses(irClass, ktClassOrObject)
+                if (!isSealedInlineSubclassOfSealedInlineClass) {
+                    generateAdditionalMembersForSingleFieldValueClasses(irClass, ktClassOrObject)
+                }
             }
 
             if (irClass.isData && ktClassOrObject is KtClassOrObject) {
                 generateAdditionalMembersForDataClass(irClass, ktClassOrObject)
             }
 
-            if (irClass.isMultiFieldValueClass && ktClassOrObject is KtClassOrObject) {
+            if (classDescriptor.isValue && classDescriptor.annotations.hasAnnotation(JVM_INLINE_ANNOTATION_FQ_NAME) &&
+                !classDescriptor.isInlineClass() && ktClassOrObject is KtClassOrObject
+            ) {
                 generateAdditionalMembersForMultiFieldValueClasses(irClass, ktClassOrObject)
             }
 
             if (DescriptorUtils.isEnumClass(classDescriptor)) {
                 generateAdditionalMembersForEnumClass(irClass)
             }
 
+            if (classDescriptor.isInline && !classDescriptor.annotations.hasAnnotation(JVM_INLINE_ANNOTATION_FQ_NAME)) {
```

Comment:
It feels weird that we _add_ an annotation which is not there on the declaration site, to supposedly fix the lack of information of whether the class was value/inline in IR... or something? And because of that, a lot of tests depend on stdlib now. Can we invent some other approach, like adding some new information to IR nodes? Maybe you can explain the problem?